### PR TITLE
Initial functionality

### PR DESCRIPTION
**Pull request by plally**

### CODE REVIEW

File: lua/autorun/server/iris_init.lua [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        userData[util.SteamIDTo64(k)] = v
+    end
+
+    postJson( "/ranks/bulk_update", {
+        users = userData,
+        realm = "cfc3",
+    }, nil, logError )
+end
+
+hook.Add( "InitPostEntity", "CFC_IrisInterface_BulkRankUpdate", sendGroupsToIris )
```
Comment (plally) [inline lua/autorun/server/iris_init.lua:48]: need to test to see if ULib.ucl.users is loaded when this is called

File: lua/autorun/server/iris_init.lua [hunk 2]
```diff
@@ -0,0 +1,48 @@
+local baseURL = "http://iris.cfcservers.org/api"
```
Comment (sarahsturgeon) [inline lua/autorun/server/iris_init.lua]:
```suggestion
local baseURL = "https://iris.cfcservers.org/api/"
```

File: lua/autorun/server/iris_init.lua [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+local authToken = file.Read( "cfc/iris/auth_token.txt" )
+authToken = string.Replace(authToken, "\n", "")
```
Comment (sarahsturgeon) [inline lua/autorun/server/iris_init.lua]:
```suggestion
authToken = string.Replace( authToken, "\n", "" )
authToken = string.Replace( authToken, "\r", "" )
```

File: lua/autorun/server/iris_init.lua [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+local function postJson( endpoint, data, callbackSuccess, callbackFailure )
+    HTTP{
+        failed = callbackFailure,
+        success = callbackSuccess,
+        method = "POST",
+        url = baseURL .. endpoint,
+        headers = {
+            "Authentication": "Bearer "..authToken
```
Comment (sarahsturgeon) [inline lua/autorun/server/iris_init.lua]:
```suggestion
            "Authorization": "Bearer " .. authToken
```

File: lua/autorun/server/iris_init.lua (hunk 1 above)
Comment (sarahsturgeon) [reply lua/autorun/server/iris_init.lua:48]: According to the wiki,
> `GM:InitPostEntity` - Good hook to create new entities. (You're still able to spawn them, but worldspawn won't be valid)

Makes me think it's fine

File: lua/autorun/server/iris_init.lua [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+end
+
+local function sendGroupsToIris()
+    local userData = {}
+
+    for k, v in pairs( ULib.ucl.users ) do
+        userData[util.SteamIDTo64(k)] = v
+    end
+
+    postJson( "/ranks/bulk_update", {
```
Comment (sarahsturgeon) [inline lua/autorun/server/iris_init.lua]:
```suggestion
    postJson( "ranks/bulk_update", {
```

File: lua/autorun/server/iris_init.lua [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+local function logError( err )
+    print( "CFC Iris Interface HTTP error: " .. err)
+end
+
+local function sendGroupsToIris()
+    local userData = {}
+
+    for k, v in pairs( ULib.ucl.users ) do
+        userData[util.SteamIDTo64(k)] = v
```
Comment (sarahsturgeon) [inline lua/autorun/server/iris_init.lua]:
```suggestion
        userData[util.SteamIDTo64( k )] = v
```

File: lua/autorun/server/iris_init.lua [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    end
+
+    postJson( "/ranks/bulk_update", {
+        users = userData,
+        realm = "cfc3",
```
Comment (sarahsturgeon) [inline lua/autorun/server/iris_init.lua]: Let's load the realm from `data/cfc/realm.txt`

File: lua/autorun/server/iris_init.lua [hunk 8]
```diff
@@ -0,0 +1,48 @@
+local baseURL = "http://iris.cfcservers.org/api"
+local authToken = file.Read( "cfc/iris/auth_token.txt" )
+authToken = string.Replace(authToken, "\n", "")
+
+local function postJson( endpoint, data, callbackSuccess, callbackFailure )
+    HTTP{
+        failed = callbackFailure,
+        success = callbackSuccess,
```
Comment (sarahsturgeon) [inline lua/autorun/server/iris_init.lua]:
```suggestion
        failed = callbackFailure or function( ... ) logger:info( ... ) end,
        success = callbackSuccess or function( ... ) logger:error( ... ) end,
```

**Review by sarahsturgeon — CHANGES_REQUESTED** [review]
A few comments, mostly nits

Explanation:

On the topic of building the URL, I think it's important to make sure that, when accepting user input, we make sure we close off the scope we want.

i.e., if we did `"https://iris.cfcservers.org/api" .. endpoint`, we'd _technically_ allow someone to pass in like `-admin/getall`. Of course it's not an actual concern in this project, but it's good practice to make sure that when building a URL, there's no chance of them escaping the scope you want them in.

We should add Logger as a requirement given that this is glue between two things we control.

Realm shouldn't be hard coded. While it kind of sucks, we already have the realm in a data file, we can just load it from there (`cfc_gmod_env`? 🤔).

While probably not actually an issue, we need to replace both `\n` and `\r` from files loaded in gmod.

**Review by sarahsturgeon — APPROVED** [review]

**Pull request merged**